app/src/main/kotlin/com/github/gotify/Repository.kt

```diff
@@ -0,0 +1,320 @@
+     * whenever the list changes.
+     */
+    val applicationsState: StateFlow<List<ApplicationState>> =
+        combine(applications, messages, paging) { apps, messages, paging ->
+            apps.map { app ->
+                val unreadCount = messages[app.id]?.size ?: 0
+                val hasMoreMessages = paging[app.id]?.hasMore ?: true
+                ApplicationState(
+                    app,
+                    unreadCount,
+                    hasMoreMessages,
+                    app.image?.let {
+                        (
+                            it.toHttpUrlOrNull() ?: baseUrl.toHttpUrlOrNull()?.newBuilder()
+                                ?.addPathSegment(it)?.build()?.toString()
+                            )?.toString()
+                    }
+                )
+            }.also { println("JcLog: ${this@Repository} number of applications -> ${it.size}") }
+        }.stateIn(scope = scope, started = SharingStarted.Lazily, initialValue = emptyList())
+
+    /**
+     * Get all messages of all applications.
+     * It returns a [StateFlow] object that emits the list of all messages whenever the list changes.
+     */
+    val allMessages: StateFlow<List<Message>> =
+        messages.map { it.values.flatten() }.distinctUntilChanged()
+            .stateIn(scope = scope, started = SharingStarted.Lazily, initialValue = emptyList())
+
+    /**
+     * Get messages of an application.
+     * It returns a [StateFlow] object that emits the list of messages of the application
+     * whenever the list changes.
+     *
+     * @param application The application.
+     *
+     * @return The list of messages of the application.
+     */
+    fun getMessages(application: Application): StateFlow<List<Message>> =
+        messages.map { it[application.id] ?: emptyList() }.distinctUntilChanged()
+            .stateIn(scope = scope, started = SharingStarted.Lazily, initialValue = emptyList())
+
+    /**
+     * Fetch all applications.
+     *
+     * @return The list of the fetched applications.
+     */
+    suspend fun fetchApps(): List<Application> =
+        applicationApi.getApps().awaitResponse().takeIf { it.isSuccessful }?.body()
+            ?.also { applications.value = it } ?: emptyList()
+
+    /**
+     * Fetch messages from an application.
+     *
+     * @param application The application.
+     * @param limit The number of messages to fetch.
+     *
+     * @return The list of the fetched messages.
+     */
+    suspend fun fetchMessages(application: Application, limit: Int = LIMIT): List<Message> {
+        val currentPagingState = paging.value[application.id] ?: PagingState()
+        return when (currentPagingState.hasMore) {
+            false -> emptyList()
+            true -> messageApi.getAppMessages(application.id, limit, currentPagingState.since)
+                .awaitResponse().takeIf { it.isSuccessful }?.body()
+                ?.also { paging.value += (application.id to it.paging.toPagingState()) }?.also {
```

Comment:
Most of the .also in this class make this hard to read, especially if it also nested in also. I'd prefer putting these into variables and then using them without nesting.